metrics/release.go

```diff
@@ -17,128 +17,197 @@ limitations under the License.
 	)
 )
 
-// RegisterCompletedRelease decrements the 'release_attempt_concurrent_total' metric, increments `release_attempt_total`
-// and registers a new observation for 'release_attempt_duration_seconds' with the elapsed time from the moment the
-// Release attempt started (Release marked as 'Running').
-func RegisterCompletedRelease(reason, strategy, target string, startTime, completionTime *metav1.Time, succeeded bool) {
-	labels := prometheus.Labels{
-		"reason":    reason,
-		"strategy":  strategy,
-		"succeeded": strconv.FormatBool(succeeded),
-		"target":    target,
+func RegisterCompletedRelease(release *v1alpha1.Release) {
+	if !release.IsReleased() {
+		return
 	}
 
+	labels := prometheus.Labels{
+		"deployment_reason":   "",
+		"post_actions_reason": "",
+		"processing_reason":   "",
+		"release_reason":      "",
+		"release_strategy":    release.Status.Processing.ReleaseStrategy,
+		"target":              release.Status.Target,
+		"validation_reason":   "",
+	}
 	ReleaseAttemptConcurrentTotal.Dec()
-	ReleaseAttemptDurationSeconds.With(labels).Observe(completionTime.Sub(startTime.Time).Seconds())
+	ReleaseAttemptDurationSeconds.
+		With(labels).
+		Observe(release.Status.CompletionTime.Sub(release.Status.StartTime.Time).Seconds())
 	ReleaseAttemptTotal.With(labels).Inc()
 }
 
-// RegisterDeployedRelease increments the 'release_attempt_deployment_total' and registers a new observation for
-// 'release_attempt_deployment_seconds' with the elapsed time from the moment the SnapshotEnvironmentBinding was
-// created to when it was marked as deployed.
-func RegisterDeployedRelease(reason, target, succeeded string, startTime, completionTime *metav1.Time) {
-	labels := prometheus.Labels{
-		"reason":    reason,
-		"succeeded": succeeded,
-		"target":    target,
+func RegisterCompletedReleaseDeployment(release *v1alpha1.Release) {
+	if !release.HasDeploymentFinished() {
+		return
 	}
 
-	ReleaseAttemptDeploymentSeconds.Observe(completionTime.Sub(startTime.Time).Seconds())
-	ReleaseAttemptDeploymentTotal.With(labels).Inc()
+	ReleaseAttemptDeploymentDurationSeconds.
+		With(prometheus.Labels{
+			"reason": "",
+			"target": release.Status.Target,
+		}).
+		Observe(release.Status.Deployment.CompletionTime.Sub(release.Status.Deployment.StartTime.Time).Seconds())
+	ReleaseAttemptConcurrentDeploymentsTotal.Dec()
 }
 
-// RegisterInvalidRelease increments the 'release_attempt_invalid_total' and `release_attempt_total` metrics.
-func RegisterInvalidRelease(reason string) {
-	ReleaseAttemptInvalidTotal.With(prometheus.Labels{"reason": reason}).Inc()
-	ReleaseAttemptTotal.With(prometheus.Labels{
-		"reason":    reason,
-		"strategy":  "",
-		"succeeded": "false",
-		"target":    "",
-	}).Inc()
+func RegisterCompletedReleaseProcessing(release *v1alpha1.Release) {
+	if !release.HasProcessingFinished() {
+		return
+	}
+
+	ReleaseAttemptProcessingDurationSeconds.
+		With(prometheus.Labels{
+			"reason":           "",
```

Comment:
why is reason blank for all of these? If we are never providing a reason, why is it even here?

Reply:
Fixed.